signer/src/storage/postgres.rs

```diff
@@ -2618,6 +2670,247 @@ impl super::DbWrite for PgStore {
 
         Ok(())
     }
+
+    async fn write_bitcoin_tx_sighash(
+        &self,
+        sighash: &model::BitcoinTxSigHash,
+    ) -> Result<(), Error> {
+        sqlx::query(
+            r#"
+                INSERT INTO sbtc_signer.bitcoin_txs_sighashes (
+                    txid
+                  , chain_tip
+                  , prevout_txid
+                  , prevout_output_index
+                  , sighash
+                  , prevout_type
+                  , validation_result
+                  , is_valid_tx
+                  , construction_version
+                )
+                VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
+                ON CONFLICT DO NOTHING;
+                "#,
+        )
+        .bind(sighash.txid)
+        .bind(sighash.chain_tip)
+        .bind(sighash.prevout_txid)
+        .bind(i32::try_from(sighash.prevout_output_index).map_err(Error::ConversionDatabaseInt)?)
+        .bind(sighash.sighash)
+        .bind(sighash.prevout_type)
+        .bind(sighash.validation_result)
+        .bind(sighash.is_valid_tx)
+        .bind(sighash.construction_version)
+        .execute(&self.0)
+        .await
+        .map_err(Error::SqlxQuery)?;
+
+        Ok(())
+    }
+
+    async fn write_bitcoin_txs_sighashes(
+        &self,
+        sighashes: Vec<model::BitcoinTxSigHash>,
+    ) -> Result<(), Error> {
+        if sighashes.is_empty() {
+            return Ok(());
+        }
+
+        let mut txid = Vec::with_capacity(sighashes.len());
+        let mut chain_tip = Vec::with_capacity(sighashes.len());
+        let mut prevout_txid = Vec::with_capacity(sighashes.len());
+        let mut prevout_output_index = Vec::with_capacity(sighashes.len());
+        let mut sighash = Vec::with_capacity(sighashes.len());
+        let mut prevout_type = Vec::with_capacity(sighashes.len());
+        let mut validation_result = Vec::with_capacity(sighashes.len());
+        let mut is_valid_tx = Vec::with_capacity(sighashes.len());
+        let mut construction_version = Vec::with_capacity(sighashes.len());
+
+        for tx_sighash in sighashes {
```

Comment:
Just a note, nothing to change for now -- this works and does achieve atomicity, but it makes for a pretty convoluted implementation (both rust and sql). I think we should opt for readability/maintainability and embrace transactions instead.

Reply:
Yeah, I am not a bit fan of the implementation either. But I thought of keeping it consistent with the other similar methods

Reply:
Man, good points on all sides here. I don't like our messy existing implementations either (and I added them 😞), and I think we should try to be consistent too.

Slight note that I've been wary of using sqlx transactions because I got bit by a bug in an earlier version of the library. It's probably fixed by now, but that was part of the why I've (perhaps mistakenly) tried not to use them.

Reply:
Transactions are such a fundamental/primary concept/feature of rdbms that if we don't trust the library to properly handle them then I wouldn't trust anything else in it tbh 😅

Reply:
I haven't seen any issues with using their `PgPool` executor though, and I haven't seen that transaction bug recently. So I don't think that we should switch libraries, at least not now.

But, that's a good point. I was just explaining why I shied away from them.

Reply:
I'll just throw two alternative implementations for you guys to consider! Personally, I lean slightly toward the first one for its simplicity and readability.  It is theorerically slower, but I can't imagine a scenario where we'd need to handle hundreds of deposits or withdrawals in a single batch, so the performance difference should be negligible... The second one is still a bit weird, but maybe less that the current implementation?

```
    async fn write_bitcoin_withdrawals_outputs(
        &self,
        withdrawal_outputs: &[model::BitcoinWithdrawalOutput],
    ) -> Result<(), Error> {
        if withdrawal_outputs.is_empty() {
            return Ok(());
        }

        let mut transaction = self.0.begin().await.map_err(Error::SqlxQuery)?;

        for output in withdrawal_outputs {
            sqlx::query(
                r#"
                INSERT INTO sbtc_signer.bitcoin_withdrawals_outputs (
                    bitcoin_txid
                  , bitcoin_chain_tip
                  , output_index
                  , request_id
                  , stacks_txid
                  , stacks_block_hash
                  , validation_result
                  , is_valid_tx
                )
                VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
                ON CONFLICT DO NOTHING
                "#,
            )
            .bind(output.bitcoin_txid)
            .bind(output.bitcoin_chain_tip)
            .bind(i32::try_from(output.output_index).map_err(Error::ConversionDatabaseInt)?)
            .bind(i64::try_from(output.request_id).map_err(Error::ConversionDatabaseInt)?)
            .bind(output.stacks_txid)
            .bind(output.stacks_block_hash)
            .bind(output.validation_result)
            .bind(output.is_valid_tx)
            .execute(&mut *transaction)
            .await
            .map_err(Error::SqlxQuery)?;
        }
        transaction.commit().await.map_err(Error::SqlxQuery)?;

        Ok(())
    }
```

```
    async fn write_bitcoin_withdrawals_outputs(
        &self,
        withdrawal_outputs: &[model::BitcoinWithdrawalOutput],
    ) -> Result<(), Error> {
        if withdrawal_outputs.is_empty() {
            return Ok(());
        }

        // Build the VALUES part of the query dynamically
        let mut param_count = 1;
        let values: Vec<String> = withdrawal_outputs
            .iter()
            .map(|_| {
                let value = format!(
                    "(${},${},${},${},${},${},${},${})",
                    param_count,
                    param_count + 1,
                    param_count + 2,
                    param_count + 3,
                    param_count + 4,
                    param_count + 5,
                    param_count + 6,
                    param_count + 7
                );
                param_count += 8;
                value
            })
            .collect();

        let query = format!(
            r#"
            INSERT INTO sbtc_signer.bitcoin_withdrawals_outputs (
                bitcoin_txid,
                bitcoin_chain_tip,
                output_index,
                request_id,
                stacks_txid,
                stacks_block_hash,
                validation_result,
                is_valid_tx
            )
            VALUES {}
            ON CONFLICT DO NOTHING
            "#,
            values.join(",")
        );

        // Build query parameters
        let mut query = sqlx::query(&query);
        for output in withdrawal_outputs {
            query = query
                .bind(output.bitcoin_txid)
                .bind(output.bitcoin_chain_tip)
                .bind(i32::try_from(output.output_index).map_err(Error::ConversionDatabaseInt)?)
                .bind(i64::try_from(output.request_id).map_err(Error::ConversionDatabaseInt)?)
                .bind(output.stacks_txid)
                .bind(output.stacks_block_hash)
                .bind(output.validation_result)
                .bind(output.is_valid_tx);
        }
        query.execute(&self.0).await.map_err(Error::SqlxQuery)?;

        Ok(())
    }
```

Reply:
Between the two I'd go with the first one. I'd default to not using a transaction too for the reasons stated above.

Reply:
The first one will be much more performant than the second one, a lot of write overhead comes from the begin/commit and index statistics updates, which when batching is limited to once if you use a transaction, otherwise each individual statement has its own begin/commit.

Reply:
But we can do this later :)

Reply:
Yeah sure, and we can revamp the old ones too.